File: cpp/base/src/main/vsdk/toolkit/media/FourierShapeDescriptor.cpp
#include "vsdk/toolkit/media/FourierShapeDescriptor.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/common/VSDK.h"
#include <cmath>
#include <cstdio>

FourierShapeDescriptor::FourierShapeDescriptor(const java::String* label) :
    ShapeDescriptor(label) {
    featureVector = new double[numberOfElements * numberOfHarmonics];
    for (int i = 0; i < numberOfElements * numberOfHarmonics; i++) {
        featureVector[i] = 0.0;
    }
}

FourierShapeDescriptor::~FourierShapeDescriptor() {
    if (featureVector != nullptr) {
        delete[] featureVector;
        featureVector = nullptr;
    }
}

void FourierShapeDescriptor::setFeature(int sphere, int harmonic, double r, double i) {
    if (sphere < 0 || sphere >= numberOfElements || harmonic < 0 || harmonic >= numberOfHarmonics) {
        return;
    }
    double harmonicAmplitude = std::sqrt(r * r + i * i);
    featureVector[sphere * numberOfHarmonics + harmonic] = harmonicAmplitude;
}

double* FourierShapeDescriptor::getFeatureVector() const {
    if (featureVector == nullptr) {
        return nullptr;
    }
    double* copy = new double[numberOfElements * numberOfHarmonics];
    for (int i = 0; i < numberOfElements * numberOfHarmonics; i++) {
        copy[i] = featureVector[i];
    }
    return copy;
}

void FourierShapeDescriptor::setFeatureVector(double* vector) {
    if (vector == nullptr) {
        return;
    }
    for (int i = 0; i < numberOfElements * numberOfHarmonics; i++) {
        featureVector[i] = vector[i];
    }
}

java::String* FourierShapeDescriptor::toString() const {
    char fullBuffer[8192];
    int offset = 0;
    offset += snprintf(fullBuffer + offset, sizeof(fullBuffer) - offset, "SphericalHarmonics amplitudes for %d spheres and %d harmonics:\n",
        numberOfElements, numberOfHarmonics);

    for (int i = 0; i < numberOfElements * numberOfHarmonics; i++) {
        char line[256];
        std::string formatted = VSDK::formatDouble(featureVector[i]);
        offset += snprintf(fullBuffer + offset, sizeof(fullBuffer) - offset, "  - %s\n", formatted.c_str());
    }

    java::String* result = new java::String(fullBuffer);
    return result;
}

File: typescript/base/src/main/vsdk/toolkit/media/FourierShapeDescriptor.ts
import { ShapeDescriptor } from "./ShapeDescriptor.js";
import { VSDK } from "../common/VSDK.js";
import { Logger } from "../common/logging/Logger.js";
export class FourierShapeDescriptor extends ShapeDescriptor {
    private featureVector = new Float64Array(512);
    public setFeature(sphere: number, harmonic: number, r: number, i: number): void {
        if (sphere < 0 || sphere >= 32 || harmonic < 0 || harmonic >= 16) return;
        this.featureVector[sphere * 16 + harmonic] = Math.hypot(r, i);
    }
    public getFeatureVector(): Float64Array {
        return this.featureVector;
    }
    public setFeatureVector(v: Float64Array | number[]): void {
        if (v.length !== 512) {
            Logger.reportMessage(
                this,
                VSDK.ERROR,
                "setFeatureVector",
                "Trying to set featurevector from incorrectly sized data!",
            );
            return;
        }
        this.featureVector = Float64Array.from(v);
    }
    public override toString(): string {
        return `SphericalHarmonics amplitudes for 32 spheres and 16 harmonics:\n${Array.from(this.featureVector, (x) => `  - ${VSDK.formatDouble(x)}\n`).join("")}`;
    }
}
